src/unix/signal.c

```diff
@@ -167,6 +172,10 @@ static void uv__signal_handler(int signum) {
 
     if (r != -1)
       handle->caught_signals++;
+
+    if ((handle->flags & UV_SIGNAL_ONE_SHOT) &&
+        (uv__signal_first_handle(handle->signum) == NULL))
+      uv__signal_unregister_handler(signum);
```

Comment:
If you are going to call `uv__signal_unregister_handler()`, you could simply register the signal handler with `.sa_flags = SA_RESETHAND` and have the kernel take care of it for you.